### ISSUE TITLE
Case of synchronous operations

### ISSUE THREAD

**Issue by theofidry** [Tolerance/Tolerance#36]
Provided we have the following `$runner` object:

```php
$callbackRunner = new CallbackOperationRunner();
$waitStrategy = new CountLimited(
    new ExponentialBackOff(
        new SleepWaiter(),
        1
    ),
    3
);

$runner = new RetryOperationRunner(
    $callbackRunner,
    $waitStrategy
);
```

When we do:

```php
$operation = new Callback(function() {
    return $this->client->locate('80.87.18.226');
});
$this->runner->run($operation);
```

We can then access to the result via `$operation->getResult()` which is a `CallbackOperation` method and not one of the `Operation` interface.

The issue here is that `$this->runner->run($operation);` is incoherent with the `OperationRunner::run(Operation): Operation` signature that says it returns an `Operation` object. The signature implies that the runner returns a new `Operation` instance, potentially different from the was passed as an input. It makes sense if we pass an operation meant to be run asynchronously for example. But in the case of a `CallbackOperation` that we want to run synchronously to get a result, this is not ideal.

So I would rather have two different runner interfaces, one for synchronous operations and in which case the `run()` signature would be `OperationRunner::run(Operation): void`, i.e. the operation passed is the same as the one returned. And a different one for asynchronous operations with the same signature as the actual runner, for which the returned operation may be a different as the one passed to the runner.

**Comment by sroze** [comment]
That's a fair point. What I would do instead is to stop returning an operation from the operation runner, that makes almost no sense in fact. I would prefer to introduced a `RunOperation` (ie an operation that have been run) that exposes a `getResult()` method (maybe a `isSuccessful()` method too). In the case of asynchronous operations, the `getResult()` method will simply throw an exception such as a `OperationResultNotAvailable`.

**Comment by theofidry** [comment]
So how would you do it for cases where the runner was returning a new operation object like the `QueuedOperation` case?

To me it makes sense that the runner is returns *something*. What doesn't make sense is maybe returning an `Operation` systematically. Instead maybe your could return the returned value if the callback returns a value and is run synchronously, an operation when it makes sense, and `null` if there is no new operation to return or nothing and throw an exception if an error occurred or even the operation failed.

The result would be that in this case we have:

```php
$operation = new Callback(function() {
    return $this->client->locate('80.87.18.226');
});
$result = $runner->run($operation);
```

1. If has been ran synchronously
  - As the `$operation` is supposed to return something we can use `$result` right away
  - If may client failed nonetheless and thrown `AddressNotFoundException`, the runner would throw this exception
  - If is not mean to return anything, can keep going as an exception would have been thrown if something went wrong
2. If has been ran asynchronously: we're not supposed to get any `$result`. If everything has ran as expected then that's it. If an error occurred an exception is expected.

What do you think?

**Comment by sroze** [comment]
It wouldn't return an `Operation` but a `RunOperation` (in the past term of `run`, could also be named `OperationResult`).

Synchronously I disagree that we only need the result. For instance for the fallback operation runner (see #7) we obviously want the result but we might want to know if that was successfully ran or not. So, for a synchronous operation we would have:
```php
$result = $runner->run(new Callback(function() {
    return $this->client->locate('80.87.18.226');
})->getResult();
```

For asynchronous operations, we would for instance receive a `QueuedOperation`, on which we could know that it was successfully queue but also get the "message ID", or things like that:
```php
$queuedOperation = $client->run(new CommandOperation($command));

$logger->debug('Queued operation', [
    'operation' => $queuedOperation
]);
```

I think that it makes more sense :)

**Comment by theofidry** [comment]
>Synchronously I disagree that we only need the result. For instance for the fallback operation runner (see #7) we obviously want the result but we might want to know if that was successfully ran or not.

I'm no following you: if the runner ran the operation successfully in this case, we get a return value (potentially no return value), if en error occurred, we get an exception. In the case of the fallback which I quite like, I would put the logger in the decorator for example. It's the runner responsibility to handle the failing operation to provide a fallback, so looks natural to have it says when it had to switch on the fallback.

If I get your solution right, you would rather have a `OperationResult` object which can gives the result (in the case of a failed operation with a fallback runner that would be the fallback result) but still be unsuccessful. I find it confusing: how can you have a failed operation that still has a result? Yeah seems legit with the fallback case. But much less if you're using the `Runner` interface for example and have no idea of the runner type. Also you would be more likely to do something like:

```php
$result = $runner->run(new Callback(function() {
    return $this->client->locate('80.87.18.226');
})->getResult();

if (! $result->isSucessful()) {
    // handle case where the operation couldn't be run properly
}
```

Where in fact you would need to do:

```php
$result = $runner->run(new Callback(function() {
    return $this->client->locate('80.87.18.226');
})->getResult();

if (! $result->isSucessful() && $result->getValue() !== null) {
    // handle case where the operation couldn't be run properly
}
```

assuming `ResultOperation::getValue()` get the real result value.

**Comment by sroze** [comment]
This is true, the idea of simply returning the result of available sounds good to me.

**Comment by theofidry** [comment]
:+

**Comment by sroze** [comment]
Fixed by #41.

**Comment by theofidry** [comment]
Cheers
